### ISSUE TITLE
metasploit_data_models version mismatch

### ISSUE THREAD

**Issue by ZeroChaos-** [rapid7/metasploit-framework#4754]
Bundler could not find compatible versions for gem "metasploit_data_models":
  In Gemfile:
    metasploit-framework-db (>= 0) ruby depends on
      metasploit-credential (~> 0.13.8) ruby depends on
        metasploit_data_models (~> 0.22.5) ruby

    metasploit-framework-db (>= 0) ruby depends on
      metasploit_data_models (0.21.3)

This kinda says it all but I'll try to explain the situation.  In Gentoo and Pentoo, users can have multiple versions of metasploit installed side by side.  To support this, the metasploit gems are allowed to install multiple versions at a time as well.  Typically the latest release and metasploit head are installed at the same time.  Most of the gems in metasploit have deps like "~>0.22.5" which means "0.22.5=< desired version <0.23".

As such, the gems in gentoo have been "slotted" so that multiple 0.x versions can be installed at the same time, but only one of each 0.x.y version can be installed at once.

Even forgetting the designed limitations in gentoo, it appears that the semantic versioning is broken here.  The installed metasploit-credential 0.13.16 satisfies the requirements for "metasploit-credential ~>0.13.8" but requires metasploit_data_models ~>0.22.5 which then causes an issue since metasploit 4.11 directly requires metasploit_data_models 0.21.3.

I believe it would be best to drop any versions of metasploit-credential which require newer metasploit_data_models 0.22 and re-release at metasploit-credential-0.14.x; this is not the only solution, but it seems the most correct to me.

**Comment by limhoff-r7** [comment]
What commit of metasploit-framework are doing this on so we can reproduce the error?  If this was an issue on master travis-ci builds would have caught it because the bundle install on travis-ci would have seen it.

**Comment by ZeroChaos-** [comment]
this is a problem seen when trying to use both the 4.11.0 release and head on the same system.  as such, unless you test every previous release on new changes your ci won't catch issues like this.

the issues is that 4.11.0 release claims it supports metasploit_concern ~>0.13.8 but once metasploit_concern started needing mdm 0.22.x it caused the dep on metasploit_concern to become invalid.

**Comment by limhoff-r7** [comment]
From IRC, it seems like you're ignoring [`Gemfile.lock`](https://github.com/rapid7/metasploit-framework/blob/4.11.0/Gemfile.lock)  From that, we know there is a combination of gems that works for 4.11.0.

As I said in #metasploit, we don't support ignoring the Gemfile.lock because it can lead to forward floating of dependencies that we haven't tested on travis-ci.  This covers more than just the dependencies we control like metasploit-concern, and metasploit_data_models, but also rspec and factory_girl where new versions have broken some of our tests before.  Please consider using the direct versions used in the Gemfile.lock instead of ignoring it when build your package version restrictions.  We can't make guarantees about other versions even if they in theory meet the version requirements declared directly in the metasploit-framework gemspec or indirectly in dependencies.

As to [your last comment](https://github.com/rapid7/metasploit-framework/issues/4754#issuecomment-74141309).  You may want to check your names and numbers since metasploit-concern has only been released up to 0.3.0, so you're probably talking about a different gem.

**Comment by ZeroChaos-** [comment]
sorry, I mean metasploit-credential not metasploit-concern

**Comment by ZeroChaos-** [comment]
"Please consider using the direct versions used in the Gemfile.lock instead of ignoring it when build your package version restrictions. We can't make guarantees about other versions even if they in theory meet the version requirements declared directly in the metasploit-framework gemspec or indirectly in dependencies."

It looks like the metasploit package and it's associated gems are using semantic versioning.  bundler expects semantic versioning.  the deps in the Gemfile, expect semantic versioning.  To say that you knowingly broke compatibilty without changing the appropriate version component to keep everything working is...disappointing.  I do not understand why you would purposely allow improper deps.  I really do not think that asking to have the latest release and head installed at the same time is so much to ask.  do I really need to allow having every possible version of metasploit-credential installed at once just because you don't follow your own versioning scheme?

**Comment by limhoff-r7** [comment]
As for the general point of when a gem should indicate an incompatible or compatible version change in semantic versioning: metasploit-credential should not make the assumption that downstream projects directly depend on metasploit_data_models.  They may only directly depend on the metasploit-credential API, in which case it is valid for the version of metasploit-credential to indicate compatibility in its version change while depending on an incompatible version of metasploit_data_models.  If we follow your recommendation, that would indicate that compatible vs incompatible version change would have to indicate incompatibility when any gem a project depends upon has an incompatible change.  Are you proposing we adopt this process for determining whether to make the version compatible or incompatible?

**Comment by ZeroChaos-** [comment]
The separation of metasploit into many pieces has both good and bad points, but realistically, all of these components *together* make up metasploit, and on their own, they are mostly nothing.

Before components of metasploit were split into a bunch of gems, things were released together, and kept in sync together.  Now things are harder, but I believe it gets easier to actually observe the connections since this is really one project with many related pieces.

tl;dr, yes, i think that it matters.  as a major benefit, doing this right means that bundler update works, *AND* you can get fixes to users much more easily since they would be able to actually update their gems.

to be entirely frank, if you followed this style of versioning properly, you wouldn't even need to ship Gemfile.lock and again, getting updates and fixes to your users is easier.

**Comment by limhoff-r7** [comment]
We'll adopt the change that an incompatible version change, meaning increment MINOR in 0.MINOR.PATCH or MAJOR in MAJOR.MINOR.PATCH when MAJOR > 0, will indicate either the public API is incompatible or a direct dependency's version indicates an incompatible version change, but we won't go so far as auditing indirect dependencies under the assumption that direct dependencies are properly semantically versioning.  This does mean that you may still have issues with gem outside the metasploit* namespace @ZeroChaos-, but that's the other gem's problem.  Not all gems in ruby are semantically versions, most notably the rails ecosystem and metasploit-framework itself, which both using marketing versions.

Also, make sure your system differentiates between pre-1.0.0 and 1.0.0+ semantic versions.  They will work differently and we reserve the right to bump to 1.0.0 on any pre-1.0.0 gem in the ecosystem whenever we get tired of being pre-1.0.0.  We'll still follow semantic versioning, we'll just get PATCH to to indicate bug fixes and MINOR for compatible changes.  The ~> will change from ~> MAJOR.MINOR.PATCH to ~> MAJOR.MINOR then.

**Comment by ZeroChaos-** [comment]
that seems to follow in line with what I was trying to suggest moving forward.  thanks for your consideration of this issue.

that said, what do we do about the current compatibility break in metasploit-credential-0.13.x?

**Comment by limhoff-r7** [comment]
All this being said.  We can't retag 4.11 or yank the currently released gems in the ecosystem, so you're just going to have not ship it in pentoo or use Gemfile.lock for 4.11.  The switch to the new meaning of incompatibility as a transitive over direct dependencies will percolate through as the gems are updated, so expect it sometime after 4.11.1.

**Comment by ZeroChaos-** [comment]
it would be possible (and it's happened before) to pull the incompatible 0.13.x releases and re-release them as 0.14.x, this would solve the issue and keep 4.11.0 working

**Comment by limhoff-r7** [comment]
@ZeroChaos- the community as a whole outside of your use case has downloaded those versions 100s to 1000s of times depending on the specific version.  metasploit-credential 013.9 has been out for over a month.  We did a yank within a day of a version being released and barely anyone had seen it.   We will fix the future, not the past versions.

**Comment by ZeroChaos-** [comment]
with this in mind, would it be possible to make a metasploit-credential 0.14 release in the near future? that would make the dep structure much easier to fix (for me at least).

**Comment by limhoff-r7** [comment]
We will release a new version of metasploit-credential when there is need for a code change in metasploit-credential or a change percolating through one of its dependencies.  If you want to speed that along, open a PR to fix something from the [Pull Review of master](https://www.pullreview.com/github/rapid7/metasploit-credential/reviews/master).

**Comment by ZeroChaos-** [comment]
very well, my in place hack will have to do for the present issue.  thanks for your consideration.

**Comment by d1rtd0g** [comment]
Some gems where installed sucessfully until recieved the following error:
Gem::InstallError: metasploit_data_models requires Ruby version >= 2.1.
An error occurred while installing metasploit_data_models (0.23.0), and Bundler
cannot continue.
Make sure that `gem install metasploit_data_models -v '0.23.0'` succeeds before
bundling.

What should I do now?

**Comment by kernelsmith** [comment]
Umm, install Ruby > 2.1?

-Josh
